### PR TITLE
Merry as type

### PR DESCRIPTION
**Pull request by erismai**
# Summary
* Append the original error message into the casted new error
* Modify the return types to (Error, bool) to indicate whether casting succeeds or not
* Modify AsType() to deal with non-merry original errors
* Add tests for nil input, non-merry original error, key-value pair assertion

# Story
* https://myhelix.atlassian.net/browse/HPLT-37

### CODE REVIEW

File: errors.go [hunk 1]
```diff
@@ -152,6 +152,34 @@ func ReplaceNativeMessage(e error, msg string) Error {
 	}
 }
 
+// AsType "casts" an input error as a target merry.Error such that the returned merry Error evaluates
+// to true when operating Is(newErr, targetErr). If e is nil, return nil. If targetType is nil,
+// A wrapped merry Error of e is returned. The key value pairs of the targetType is kept and then
+// appended with key value map from the input error. Stack trace of the original input error is maintained.
+func AsType(e error, targetType Error) (Error, bool) {
```
Comment (wangdrew) [inline errors.go]: I don't feel it's necessary to return a success bool as the 2nd return parameter. If the client calls this function with nil for either or both arguments, it should expect either a nil or the same merry wrapped error. Garbage in - garbage out. Adding a 2nd parameter forces all clients of this function to have to handle the case where the AsType "fails" and I don't really see the usefulness of this.

File: errors_test.go [hunk 2]
```diff
@@ -521,6 +520,127 @@ func TestVerboseDefault(t *testing.T) {
...
+
+	// newMerryErr is nil
+	assert.Equal(t, nil, nilOrigin)
+	assert.Equal(t, nil, newMerryErr)
+
+	// newMerryErr and nilOrigin are both nil, thus Is(newMerryErr, nilOrigin) return true
+	// But if only one of the two params passed in Is() is nil, it returns false
+	assert.True(t, Is(newMerryErr, nilOrigin))
+	assert.False(t, Is(newMerryErr, castType))
+	assert.False(t, ok)
```
Comment (wangdrew) [inline errors_test.go]: nit: remove duplicate assertion, `ok` is already being asserted in line 559

File: errors_test.go [hunk 3]
```diff
@@ -521,6 +520,127 @@ func TestVerboseDefault(t *testing.T) {
...
+	// test new merry error is castType, instead of origin's type
+	assert.True(t, Is(newMerryErr, castType))
+	assert.False(t, Is(newMerryErr, origin))
+
+	// test new error message is <castType error message>: <original error message>
+	assert.Equal(t, castType.Error()+": "+origin.Error(), newMerryErr.Error())
+	assert.NotEqual(t, origin.Error(), newMerryErr.Error())
+
+	// newMerryErr is not equal of original error any more but has the same stack trace as origin
+	assert.Equal(t, Stacktrace(newMerryErr), originStackTrace)
```
Comment (wangdrew) [inline errors_test.go]: nice 👍

File: errors_test.go [hunk 4]
```diff
@@ -521,6 +520,127 @@ func TestVerboseDefault(t *testing.T) {
...
+	castType = WithValue(castType, "foo", "bar2")
+	newMerryErr, ok = AsType(origin, castType)
+
+	// test new merry error is castType, instead of origin's type
+	assert.True(t, Is(newMerryErr, castType))
+	assert.False(t, Is(newMerryErr, origin))
+	assert.True(t, ok)
+
+	// Value(newMerryErr, "foo") should be the same as origin
+	assert.Equal(t, "bar1", Value(newMerryErr, "foo"))
```
Comment (wangdrew) [inline errors_test.go]: nice 👍

**Review by wangdrew — CHANGES_REQUESTED** [review]
Nice job on this PR! Would like to see one change with regards to the `AsType()` function return params. Open to discussion if you think otherwise.

File: errors.go (hunk 1 above)
Comment (MayaLi) [reply errors.go]: Having the second parameter is akin to doing `w, ok := e.(*merryErr)` in which the client is notified that the casting worked or not. 

For the case that `targetType` is nil, it is unclear what the intended behavior should be. The client expects the returned `Error` to be of `targetType`, instead of asking the client to do a `Is()` comparison, we can simplify return `false` to indicate that the cast did not work as wished. This is the same idea with the case of input `e` being nil.

Comment (erismai) [reply errors.go]: Either of the design looks good to me. It all depends on what the client expect it to be. @alexherns seems to prefer just returning nil?

Comment (alexherns) [reply errors.go]: @MayaLi -- to me the difference between this function and type casting is that in golang, handling the `ok` return argument is optional and by doing so breaks with almost every other convention in the language. I.e. both of these statements are legal:
```go
w, ok := e.(*merryErr)
w := e.(*merryErr)
```
For better or worse, we aren't allowed to write functions in that way, which means that, as @wangdrew suggests, every consumer would have to at least receive the 2nd return parameter. And sure, all consumers _could_ just do `err, _ = merry.AsType(err, OtherErrorType)`, but I think that having the 2nd return parameter there would give me the feeling that I should do something with it, which adds additional complexity to my code every time I call `merry.AsType`.

File: errors.go [hunk 5]
```diff
@@ -152,6 +152,34 @@ func ReplaceNativeMessage(e error, msg string) Error {
...
+
+		inputMap := Values(e)
+		for k, v := range inputMap {
+			output = output.WithValue(k, v)
+		}
+		return Append(output, e.Error()), true
+
+	default:
+		output := targetType.Here()
+		return Append(output, e.Error()), true
```
Comment (alexherns) [inline errors.go]: If not a merry error we should probably use `ReplaceNativeMessage` instead, otherwise the errors can be reported with all different messages depending on the value reported by the underlying reporter (@erismai -- ask one of us if this is unclear)

File: errors.go (hunk 1 above)
Comment (chriswhelix) [reply errors.go]: Two return args also means you can't use chaining, which is a pattern generally supported by merry.

File: errors.go (hunk 5 above)
Comment (chriswhelix) [reply errors.go]: My feeling is it should always be the user's responsibility to determine where ReplaceNativeMessage is needed. Not all projects (even all Helix projects) may be using Rollbar.

File: errors.go [hunk 6]
```diff
@@ -152,6 +152,34 @@ func ReplaceNativeMessage(e error, msg string) Error {
...
+
+	case *merryErr:
+		output := targetType.Here()
+		output = output.WithValue(stack, Value(e, stack))
+
+		inputMap := Values(e)
+		for k, v := range inputMap {
+			output = output.WithValue(k, v)
+		}
+		return Append(output, e.Error()), true
```
Comment (chriswhelix) [inline errors.go]: I'd like to propose an alternate approach; rather than reconstructing a whole new error chain and trying to copy all the relevant info over, what about adding explicit support for multiple identities to .Is?

In other words, implementation would be like:

```go
// This is the pattern used by userMessage, httpCode, etc. 
const additionalErrorIdentities = "additional error identities"

func (e *merryErr) WithErrorIdentity(err error) Error {
   var idents []error
    if i, ok := e.Value(additionalErrorIdentities).([]error) {
        idents = i
    }
    append(idents, err)
    return e.WithValue(additionalErrorIdentities, idents)
}
```

Then .Is would know how to check additionalErrorIdentities.

File: errors.go (hunk 1 above)
Comment (MayaLi) [reply errors.go]: Chaining is a good reason for a single parameter return. I just read merry code again and looks like the functions account for nil input. That's safe enough.

**Review by chriswhelix — APPROVED** [review]
👍 nice! There's a weird note at the end of the diff about an added submodule. What's that about?

**Review by MayaLi — COMMENTED** [review]
Good changes. While it is implied that it should behave this way, but I think you should add test cases to demonstrate that the stacktrace and values attached to the error are maintained after calling `WithErrorIdentity`. It is also not clear to me that the values attached to the new error identity are maintained, which in my mind is a desired behavior

**Review by wangdrew — APPROVED** [review]
LGTM nice work figuring out the recursive logic!

**Review by alexherns — APPROVED** [review]
